### PR TITLE
#14 Implement MMAC metric calculation

### PR DESCRIPTION
**Pull request by sergey-karazhenets**

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256 please, pay attention to this pull request

**Comment by 0crat** [comment]
Job `gh:yegor256/jpeek#41` is in scope.

**Comment by yegor256** [comment]
@memoyil can you please review this PR?

**Comment by memoyil** [comment]
Ok i will do it

**Comment by yegor256** [comment]
@sergey-karazhenets you didn't update the README page, please do it

File: src/main/java/org/jpeek/metrics/cohesion/MMAC.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            int mcnt = 0;
+            for (final Collection<String> method : methods) {
+                if (method.contains(type)) {
+                    ++mcnt;
+                }
+            }
+            sum += mcnt * (mcnt - 1);
+        }
+        return sum
+            / (types.size() * methods.size() * (methods.size() - 1));
```
Comment (memoyil) [inline src/main/java/org/jpeek/metrics/cohesion/MMAC.java]: I dont think this is correct, it should be `1- (2/(methods.size() * (methods.size() - 1))*sum`

Comment (sergey-karazhenets) [reply src/main/java/org/jpeek/metrics/cohesion/MMAC.java]: Hmm, you partially right. Original formula for _MMAC_ is:
mmac

See please [this one doc](http://waset.org/publications/5239/a-design-based-cohesion-metric-for-object-oriented-classes), chapter IV, paragraph "A. MMAC and AAC Metrics".

In your formula `sum` is sum of normalized distances between pair of methods, but I don't calculate normalized distances to `sum` variable. I calculate sum of methods which use the same types of parameters or return types and then divide this _sum_ to multiplication of distinct types and methods. It is another analogous way to calculate _MMAC_ metric. And it is described in mentioned above doc. See please formula (7) in doc
mmac_7
 and how we can get this one formula from original formula (original formula is presented as (2) in doc).

Comment (memoyil) [reply src/main/java/org/jpeek/metrics/cohesion/MMAC.java]: Correct, thank you. 

@yegor256 You can merge this pr.

**Comment by yegor256** [comment]
@sergey-karazhenets please, add the PDF file to the `papers` folder and I will merge it

**Comment by sergey-karazhenets** [comment]
@yegor256 I've added PDF to papers

**Comment by yegor256** [comment]
@rultor merge

**Pull request merged**